### PR TITLE
Clone VM

### PR DESCRIPTION
**Pull request by avivtur**
## 📝 Description

Adding a clone VM modal and action

## 🎥 Demo 

(updated**)

https://user-images.githubusercontent.com/67270715/162612176-a7bfe2db-5b12-4a7c-9415-d8aa4a4368b1.mp4

Signed-off-by: Aviv Turgeman <aturgema@redhat.com>

### CODE REVIEW

File: src/views/virtualmachines/actions/components/CloneVMModal/hooks/useProjectsAndDVsAndPVCs.tsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+type UseProjectsAndDVsAndPVCs = (vm: V1VirtualMachine) => {
+  projects: K8sResourceCommon[];
+  dataVolumes: V1beta1DataVolume[];
+  pvcs: IoK8sApiCoreV1PersistentVolumeClaim[];
+  loaded: boolean;
+  error: any;
+};
+
+const useProjectsAndDVsAndPVCs: UseProjectsAndDVsAndPVCs = (vm: V1VirtualMachine) => {
+  //   const requestsDataVolumes = !!(getVolumes(vm) || []).find((volume) => volume?.dataVolume?.name);
```
Comment (glekner) [inline src/views/virtualmachines/actions/components/CloneVMModal/hooks/useProjectsAndDVsAndPVCs.tsx:24]: comments?

File: src/views/virtualmachines/actions/components/CloneVMModal/CloneVMModal.tsx [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+    return clonedVM;
+  }, [cloneDescription, cloneName, cloneProject, dataVolumes, pvcs, startCloneVM, vm]);
+
+  const onClone = (updatedVM: V1VirtualMachine) => {
+    if (isVMRunning) {
+      return stopVM(vm).then(() => k8sCreate({ model: VirtualMachineModel, data: updatedVM }));
+    }
+    return k8sCreate({ model: VirtualMachineModel, data: updatedVM });
+  };
```
Comment (glekner) [inline src/views/virtualmachines/actions/components/CloneVMModal/CloneVMModal.tsx:88]:
```suggestion
  const onClone = (updatedVM: V1VirtualMachine) => {
    const createPromise = () => k8sCreate({ model: VirtualMachineModel, data: updatedVM });
  
    if (isVMRunning) {
      return stopVM(vm).then(createPromise);
    }
    return createPromise();
  };
```

File: src/views/virtualmachines/actions/components/CloneVMModal/components/ConfigurationSummary.tsx [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+      <TextListItem component={TextListItemVariants.dd}>
+        {(getInterfaces(vm) || [])?.map(({ name, model }) => (
+          <div key={name}>{model ? `${name} - ${model}` : name}</div>
+        ))}
+      </TextListItem>
+      <TextListItem className="text-muted" component={TextListItemVariants.dt}>
+        {t('Disks')}
+      </TextListItem>
+      <TextListItem component={TextListItemVariants.dd}>
+        {getClonedDisksSummery(vm, pvcs, dataVolumes)}
```
Comment (glekner) [inline src/views/virtualmachines/actions/components/CloneVMModal/components/ConfigurationSummary.tsx:61]: Summery => Summary

File: src/views/virtualmachines/actions/components/CloneVMModal/hooks/useProjectsAndDVsAndPVCs.tsx [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+type UseProjectsAndDVsAndPVCs = (vm: V1VirtualMachine) => {
+  projects: K8sResourceCommon[];
+  dataVolumes: V1beta1DataVolume[];
+  pvcs: IoK8sApiCoreV1PersistentVolumeClaim[];
+  loaded: boolean;
+  error: any;
+};
+
+const useProjectsAndDVsAndPVCs: UseProjectsAndDVsAndPVCs = (vm: V1VirtualMachine) => {
```
Comment (glekner) [inline src/views/virtualmachines/actions/components/CloneVMModal/hooks/useProjectsAndDVsAndPVCs.tsx:23]: `useCloneVmResources`

File: src/views/virtualmachines/actions/components/CloneVMModal/hooks/useProjectsAndDVsAndPVCs.tsx [hunk 5]
```diff
@@ -0,0 +1,61 @@
+import {
+  modelToGroupVersionKind,
+  PersistentVolumeClaimModel,
+  ProjectModel,
+} from '@kubevirt-ui/kubevirt-api/console';
+import DataVolumeModel from '@kubevirt-ui/kubevirt-api/console/models/DataVolumeModel';
+import { V1beta1DataVolume } from '@kubevirt-ui/kubevirt-api/containerized-data-importer/models';
+import { IoK8sApiCoreV1PersistentVolumeClaim } from '@kubevirt-ui/kubevirt-api/kubernetes/models';
+import { V1VirtualMachine } from '@kubevirt-ui/kubevirt-api/kubevirt';
+// import { getVolumes } from '@kubevirt-utils/resources/vm';
```
Comment (glekner) [inline src/views/virtualmachines/actions/components/CloneVMModal/hooks/useProjectsAndDVsAndPVCs.tsx:10]: comment

File: src/views/virtualmachines/actions/components/CloneVMModal/utils/helpers.tsx [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+import {
+  getDataVolumeTemplates,
+  getDisks,
+  getInterfaces,
+  getVolumes,
+} from '@kubevirt-utils/resources/vm';
+import { formatBytes } from '@kubevirt-utils/resources/vm/utils/disk/size';
+import { K8sResourceCommon } from '@openshift-console/dynamic-plugin-sdk';
+
+export const getName = (obj: K8sResourceCommon) => obj?.metadata?.name;
```
Comment (glekner) [inline src/views/virtualmachines/actions/components/CloneVMModal/utils/helpers.tsx:18]: no need for these helpers, either we already have them or you can just use `obj?.metadata?.name` directly

File: src/views/virtualmachines/actions/components/CloneVMModal/utils/helpers.tsx [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+export const getNamespace = (obj: K8sResourceCommon) => obj?.metadata?.namespace;
+
+export const getRandomChars = (len: number): string => {
+  return Math.random()
+    .toString(36)
+    .replace(/[^a-z0-9]+/g, '')
+    .substr(1, len);
+};
+
+export const getClonedDisksSummery = (
```
Comment (glekner) [inline src/views/virtualmachines/actions/components/CloneVMModal/utils/helpers.tsx:29]: getClonedDisksSummary

Comment (avivtur) [reply src/views/virtualmachines/actions/components/CloneVMModal/utils/helpers.tsx]: +1 changed

File: src/views/virtualmachines/actions/components/CloneVMModal/utils/helpers.tsx (hunk 6 above)
Comment (avivtur) [reply src/views/virtualmachines/actions/components/CloneVMModal/utils/helpers.tsx]: +1 removed

File: src/views/virtualmachines/actions/components/CloneVMModal/hooks/useProjectsAndDVsAndPVCs.tsx (hunk 5 above)
Comment (avivtur) [reply src/views/virtualmachines/actions/components/CloneVMModal/hooks/useProjectsAndDVsAndPVCs.tsx]: +1 removed

File: src/views/virtualmachines/actions/components/CloneVMModal/hooks/useProjectsAndDVsAndPVCs.tsx (hunk 4 above)
Comment (avivtur) [reply src/views/virtualmachines/actions/components/CloneVMModal/hooks/useProjectsAndDVsAndPVCs.tsx]: +1 changed

File: src/views/virtualmachines/actions/components/CloneVMModal/components/ConfigurationSummary.tsx (hunk 3 above)
Comment (avivtur) [reply src/views/virtualmachines/actions/components/CloneVMModal/components/ConfigurationSummary.tsx]: +1 changed

File: src/views/virtualmachines/actions/components/CloneVMModal/CloneVMModal.tsx (hunk 2 above)
Comment (avivtur) [reply src/views/virtualmachines/actions/components/CloneVMModal/CloneVMModal.tsx]: +1 changed

File: src/views/virtualmachines/actions/components/CloneVMModal/hooks/useProjectsAndDVsAndPVCs.tsx (hunk 1 above)
Comment (avivtur) [reply src/views/virtualmachines/actions/components/CloneVMModal/hooks/useProjectsAndDVsAndPVCs.tsx]: +1 removed

**Pull request merged**